usr.sbin/cncctl/cncctl.8
.\"	$FabBSD$
.\"
.\" Copyright (c) 2009 Hypertriton, Inc. <http://hypertriton.com/>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd $Mdocdate: April 2 2009 $
.Dt CNCCTL 8
.Os
.Sh NAME
.Nm cncctl
.Nd configure the motion control (cnc) device
.Sh SYNOPSIS
.Nm cncctl
.Op Fl Cc
.Op Fl s Ar modifier
.Op Fl F Ar modifier
.Op Fl p Ar coords
.Sh DESCRIPTION
The
.Nm
utility configures the motion control system via the
.Xr cnc 4
interface.
The commands are as follows:
.Bl -tag -width Ds
.It Fl C
Calibrate the timings necessary for the trajectory planner's spin delay loops.
The operation can take several minutes and the FabBSD kernel will not respond
to interrupts during the calibration.
This only needs to be done once for any given machine.
The results will be saved to
.Pa /var/db/cnc.timings .
.It Fl c
Reread the timing calibration data from
.Pa /var/db/cnc.timings .
This command is typically invoked by the
.Pa /etc/rc
boot script.
.It Fl s Ar modifier
Show the cnc parameters specified by
.Ar modifier
(may be abbreviated):
.Pp
.Bl -tag -width xxxxxxxxxxx -compact
.It Fl s Cm devices
List the devices registered with the motion control system.
.It Fl s Cm info
Display various statistics and counters.
.It Fl s Cm limits
Display the kinematic limits in effect.
.It Fl s Cm position
Print the current machine position (in steps).
.El
.It Fl F Ar modifier
Flush the cnc parameters specified by
.Ar modifier
(may be abbreviated):
.Pp
.Bl -tag -width xxxxxxxxxxx -compact
.It Fl s Cm info
Flush all statistics and counters.
.El
.It Fl p Ar coords
Set the current machine position.
This changes the kernel's idea of the actual machine position, no motion is
performed.
The boot scripts use this command to save and restore the coordinates across
system reboots.
.El
.Sh FILES
.Bl -tag -width "/var/db/cnc.timings" -compact
.It Pa /etc/cncctl.conf
Packet filter rules file.
.It Pa /var/db/cnc.timings
Timing calibration data.
.El
.Sh SEE ALSO
.Xr cnc 4 ,
.Xr rc 8 ,
.Xr rc.conf 8
.Sh HISTORY
The
.Nm
program first appeared in FabBSD 1.0.


trunk/usr.sbin/cncctl/cncctl.c
/*	$FabBSD$	*/
/*
 * Copyright (c) 2009 Hypertriton, Inc. <http://hypertriton.com/>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * Program to control the cnc(4) interface.
 */

#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/limits.h>

#include <limits.h>
#include <cnc.h>
#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "pathnames.h"

static const char *showopt_list[] = {
	"devices", "info", "limits", "position", NULL
};
static const char *flushopt_list[] = {
	"info", NULL
};

const char *device = _PATH_DEV_CNC;
const char *timingsfile = _PATH_TIMINGS;
const char *showopt = NULL;
const char *flushopt = NULL;
int devfd = -1;
int quiet = 0;

static void
printusage(const char *optlist[])
{
	extern char *__progname;

	if (optlist != NULL) {
		const char **s = &optlist[0];
		fprintf(stderr, "acceptable modifiers: ");
		while (*s != NULL) {
			fputs(*s, stderr);
			fputc(' ', stderr);
			s++;
		}
		fputc('\n', stderr);
	} else {
		fprintf(stderr, "usage: %s [-qCc] [-c timingfile] "
		                "[-s modifier] [-F modifier] [-p coords]\n",
				__progname);
	}
	exit(1);
}

static const char *
lookup_option(char *cmd, const char **list)
{
	if (cmd != NULL && *cmd) {
		for (; *list; list++)
			if (!strncmp(cmd, *list, strlen(cmd)))
				return (*list);
	}
	return (NULL);
}

static int
calibrate_timings(void)
{
	struct cnc_timings t;
	FILE *f;

	t.hz = 0;
	t.move_jog = 0;
	
	if ((f = fopen(_PATH_TIMINGS, "r")) != NULL) {
		char buf[128], *s = &buf[0];
		char *s1, *s2;

		fread(buf, sizeof(buf), 1, f);
		s1 = strsep(&s, "\n");
		s2 = strsep(&s, "\n");
		t.hz = (cnc_utime_t)strtoull(s1, (char **)NULL, 10);
		t.move_jog = (cnc_utime_t)strtoull(s2, (char **)NULL, 10);
		fclose(f);
	}
	if (ioctl(devfd, CNC_SETTIMINGS, &t) == -1) {
		err(1, "CNC_SETTIMINGS");
	}
	if (t.hz == 0 || t.move_jog == 0) {
		printf("Calibrating timings. This may take several minutes...\n");
		if (ioctl(devfd, CNC_CALTIMINGS, &t) == -1) {
			err(1, "CNC_CALTIMINGS");
		}
		printf("Saving timings to %s\n", _PATH_TIMINGS);
		if ((f = fopen(_PATH_TIMINGS, "w")) != NULL) {
			fprintf(f, "%llu\n%llu\n", t.hz, t.move_jog);
			fclose(f);
		} else {
			warn("%s", _PATH_TIMINGS);
		}
	}
	if (!quiet) {
		printf("Timings calibrated successfully (1Hz=%llu, jog=%llu)\n",
		    t.hz, t.move_jog);
	}
	return ioctl(devfd, CNC_SETTIMINGS, &t);
}

int
main(int argc, char *argv[])
{
	struct cnc_vector pos;
	char pb[128];
	int ch;
	int do_cal = 0;
	
	if ((devfd = open(device, O_WRONLY)) == -1)
		err(1, "%s", device);

	while ((ch = getopt(argc, argv, "Cs:F:p:")) != -1) {
		switch (ch) {
		case 'C':
			do_cal = 1;
			break;
		case 's':
			if ((showopt = lookup_option(optarg, showopt_list))
			    == NULL) {
				warnx("Unknown show modifier '%s'", optarg);
				printusage(showopt_list);
			}
			break;
		case 'F':
			if ((flushopt = lookup_option(optarg, flushopt_list))
			    == NULL) {
				warnx("Unknown flush modifier '%s'", optarg);
				printusage(flushopt_list);
			}
			break;
		case 'q':
			quiet = 1;
			break;
		case 'p':
			if (!quiet) {
				if (ioctl(devfd, CNC_GETPOS, &pos) == -1) {
					errx(1, "CNC_GETPOS");
				}
				cnc_vec_print(&pos, pb, sizeof(pb));
				printf("%s -> %s\n", pb, optarg);
			}
			if (cnc_vec_parse(&pos, optarg) == -1)
				errx(1, "-p: %s", cnc_get_error());
			if (ioctl(devfd, CNC_SETPOS, &pos) == -1)
				errx(1, "CNC_SETPOS");
			break;
		default:
			printusage(NULL);
		}
	}
	if (argc == 0)
		printusage(NULL);
	
	if (do_cal) {
		calibrate_timings();
	} else if (showopt) {
		switch (*showopt) {
		case 'p':
			{
				struct cnc_vector pos;
				char pb[128];
				if (ioctl(devfd, CNC_GETPOS, &pos) == -1) {
					errx(1, "CNC_GETPOS");
				}
				cnc_vec_print(&pos, pb, sizeof(pb));
				fputs(pb, stdout);
				fputc('\n', stdout);
			}
			break;
		case 'd':
			{
				struct cnc_device_info di;
				if (ioctl(devfd, CNC_GETDEVICEINFO, &di) == -1) {
					errx(1, "CNC_GETDEVICEINFO");
				}
				printf("Devices:\n");
				printf("  %d servos\n", di.nservos);
				printf("  %d spindles\n", di.nspindles);
				printf("  %d estops\n", di.nestops);
				printf("  %d encoders\n", di.nencoders);
				printf("  %d MPGs\n", di.nmpgs);
			}
			break;
		case 'i':
			{
				struct cnc_stats st;
				if (ioctl(devfd, CNC_GETSTATS, &st) == -1) {
					errx(1, "CNC_GETSTATS");
				}
				printf("Counters\n");
				printf("  peak-velocity %lu\n",	(u_long)st.peak_velocity);
				printf("  e-stops       %lu\n",	(u_long)st.estops);
			}
			break;
		case 'l':
			{
				struct cnc_kinlimits kl;
				if (ioctl(devfd, CNC_GETKINLIMITS, &kl) == -1) {
					errx(1, "CNC_GETKINLIMITS");
				}
				printf("Kinematic limits\n");
				printf("  max-velocity %lu\n", (u_long)kl.Fmax);
				printf("  max-accel    %lu\n", (u_long)kl.Amax);
				printf("  max-jerk     %lu\n", (u_long)kl.Jmax);
			}
			break;
		}
	} else if (flushopt) {
		switch (*showopt) {
		case 'i':
			if (ioctl(devfd, CNC_CLRSTATS) == -1)
				errx(1, "CNC_CLRSTATS");
			break;
		}
	}

	close(devfd);
	return (0);
}
